File: src/lib/services/system/keychain.ts
import { invoke } from '@tauri-apps/api/core';
import { log } from '@/utils/logger';

export interface KeychainResult {
  found: boolean;
  value: string | null;
  available: boolean;
  error?: string;
}

/**
 * Stores a secret in the OS keychain.
 * Throws if the keychain is unavailable — NO plaintext fallback.
 */
export async function setSecret(key: string, value: string): Promise<void> {
  await invoke('set_secret', { key, value });
  log.info('Secret stored in keychain', { key });
}

/**
 * Reads a secret from the OS keychain.
 * Returns { found: false } when no value is stored.
 * Returns { available: false } when the keychain itself is unavailable.
 * Secrets NEVER reach any store — consume result immediately in UI.
 */
export async function getSecret(key: string): Promise<KeychainResult> {
  try {
    const value = await invoke<string | null>('get_secret', { key });
    return { found: value !== null, value, available: true };
  } catch (error) {
    const msg = String(error);
    log.warn('Keychain get failed', { key, error: msg });
    return { found: false, value: null, available: false, error: msg };
  }
}

/**
 * Deletes a secret from the OS keychain.
 */
export async function deleteSecret(key: string): Promise<void> {
  await invoke('delete_secret', { key });
  log.info('Secret deleted from keychain', { key });
}

File: src-tauri/src/commands/system/keychain.rs
//! Keychain secret storage commands.
//! Uses the system keychain (macOS Keychain, Windows Credential Manager, Linux Secret Service).
//! Secrets NEVER fall back to plaintext storage.
//!
//! Return contract (matches TypeScript KeychainResult type):
//!   - Ok(Some(v)) — secret found
//!   - Ok(None)    — no secret stored for this key (not an error)
//!   - Err(msg)    — keychain unavailable or OS permission denied

const SERVICE: &str = "bismuth-pkm";
const MAX_KEY_LEN: usize = 128;

fn validate_key(key: &str) -> Result<(), String> {
    if key.is_empty() || key.len() > MAX_KEY_LEN {
        return Err(format!("Secret key must be 1–{} characters", MAX_KEY_LEN));
    }
    Ok(())
}

/// Stores a secret in the OS keychain.
/// Returns Err if the keychain is unavailable — no plaintext fallback.
#[tauri::command]
pub async fn set_secret(key: String, value: String) -> Result<(), String> {
    validate_key(&key)?;
    keyring_set(SERVICE, &key, &value)
}

/// Reads a secret from the OS keychain.
/// Returns Ok(None) when no value is stored (not an error).
/// Returns Err when the keychain itself is unavailable.
#[tauri::command]
pub async fn get_secret(key: String) -> Result<Option<String>, String> {
    validate_key(&key)?;
    keyring_get(SERVICE, &key)
}

/// Deletes a secret from the OS keychain.
#[tauri::command]
pub async fn delete_secret(key: String) -> Result<(), String> {
    validate_key(&key)?;
    keyring_delete(SERVICE, &key)
}

// ─── Platform keyring adapter ─────────────────────────────────────────────────
// These functions wrap whatever keyring implementation is available.
// Currently stubbed pending tauri-plugin-keychain integration.
// The stubs return Ok(None) so the UI correctly shows "no secret stored".

fn keyring_set(_service: &str, _key: &str, _value: &str) -> Result<(), String> {
    tracing::info!("keyring_set: service={} key={} (stub — keychain plugin not yet integrated)", _service, _key);
    Ok(())
}

fn keyring_get(_service: &str, _key: &str) -> Result<Option<String>, String> {
    tracing::info!("keyring_get: service={} key={} (stub — keychain plugin not yet integrated)", _service, _key);
    Ok(None)
}

fn keyring_delete(_service: &str, _key: &str) -> Result<(), String> {
    tracing::info!("keyring_delete: service={} key={} (stub — keychain plugin not yet integrated)", _service, _key);
    Ok(())
}
